CHANGELOG.md

```diff
@@ -5,6 +5,10 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
```

Comment:
@rjavadi I left quite a few comments about smaller issues int he current code base, it seems the PR is in a sort of intermediate state as the style of expressions still suggests partial filtering while the application was done without them, etc

This is probably the most important thing we need to decide asap. However your recent comments have opened another question that we should clarify: We do not know the difference between streamingAPI and lazy dataframes. The fact that you encounter memory errors even with lazy dataframes is not immediately clear to us. Can you explain that in our next meeting in detail? Ive also seen a lot of online content about polars joins where its eg suggested that different join types + postfiltering are prefered and much leaner on memory, I dont understand why but it could be worth to explore. Our expectation was definitely not that the polars method would crash around the same size as the pandas one

Reply:
>  However your recent comments have opened another question that we should clarify: We do not know the difference between streamingAPI and lazy dataframes...Our expectation was definitely not that the polars method would crash around the same size as the pandas one

I explained everything I knew about them under the discussion. In the beginning of project you recommended me to use LazyFrame in order to fix some problems you had with Pandas. I'm not a Pandas/Polars expert, so I went with your suggestion. 

> that different join types + postfiltering are prefered and much leaner on memory,
I appreciate if you could share the resource. In your case, only `cross` join could yield the result.
And from what I understand, we are basically doing join + post-filtering at the moment?

Reply:
In a nutshell, the lazy API allows you to work with larger than memory datasets **using streaming API**. In cases where streaming API is not supported (e.g. join operation), we won't be able to handle larger than memory dataframes, whether lazy or not.

Reply:
did you look at [this](https://github.com/pola-rs/polars/issues/5124) and [this](https://github.com/pola-rs/polars/pull/5339) ? it seems they do cross join and do streaming
<img width="785" alt="image" src="https://github.com/emdgroup/baybe/assets/17951239/238d0673-3576-45ef-8b14-c631742a7ea9">
note the `allow_streaming` option

Where did your attempts actually fail? during filtering? during crossjoing? during collect?

Reply:
naively in the above picture, the following lines should be equivalent tos tuff in your code

1) `lf.join` - this is what `parameter_cartesian_prod_polars` should do
2) `.filter(... polars expressions of the constraints combined with &)` - this is what `_apply_polars_constraint_filter` should do
3) `.collect(allow_streaming=True)` -  this is what should be done before the pandas filters are applied in `from_product`

Reply:
Furhtermore, can you explect the query plan via `.explain` after you created the lazy frame? we should see that all operations in there are inside `STREAMING` block
eg see here
<img width="774" alt="image" src="https://github.com/emdgroup/baybe/assets/17951239/aef9baf6-f0ac-498f-ad01-633254e6c929">

Reply:
Thanks for sending links, I'm going to see if it works.

> Where did your attempts actually fail? during filtering? during crossjoing? during collect?
During Collect.

According to `explain()` output, sum and excludsion constraints are streamable, but product constraint query is not. This example is for exclusion constraint:
```
--- STREAMING
FILTER [(col("cat_1").is_in([Series])) | ([(col("disc_5")) > (50.0)])].not() FROM
CROSS JOIN:
LEFT PLAN ON: []
  CROSS JOIN:
  LEFT PLAN ON: []
    CROSS JOIN:
    LEFT PLAN ON: []
      CROSS JOIN:
      LEFT PLAN ON: []
        CROSS JOIN:
        LEFT PLAN ON: []
          CROSS JOIN:
          LEFT PLAN ON: []
            CROSS JOIN:
            LEFT PLAN ON: []
              CROSS JOIN:
              LEFT PLAN ON: []
                CROSS JOIN:
                LEFT PLAN ON: []
                  CROSS JOIN:
                  LEFT PLAN ON: []
                    DF ["disc_0"]; PROJECT */1 COLUMNS; SELECTION: "None"
                  RIGHT PLAN ON: []
                    DF ["disc_1"]; PROJECT */1 COLUMNS; SELECTION: "None"
                  END CROSS JOIN
                RIGHT PLAN ON: []
                  DF ["disc_2"]; PROJECT */1 COLUMNS; SELECTION: "None"
                END CROSS JOIN
              RIGHT PLAN ON: []
                DF ["disc_3"]; PROJECT */1 COLUMNS; SELECTION: "None"
              END CROSS JOIN
            RIGHT PLAN ON: []
              DF ["disc_4"]; PROJECT */1 COLUMNS; SELECTION: "None"
            END CROSS JOIN
          RIGHT PLAN ON: []
            DF ["disc_5"]; PROJECT */1 COLUMNS; SELECTION: "None"
          END CROSS JOIN
        RIGHT PLAN ON: []
          DF ["disc_6"]; PROJECT */1 COLUMNS; SELECTION: "None"
        END CROSS JOIN
      RIGHT PLAN ON: []
        DF ["disc_7"]; PROJECT */1 COLUMNS; SELECTION: "None"
      END CROSS JOIN
    RIGHT PLAN ON: []
      DF ["disc_8"]; PROJECT */1 COLUMNS; SELECTION: "None"
    END CROSS JOIN
  RIGHT PLAN ON: []
    DF ["disc_9"]; PROJECT */1 COLUMNS; SELECTION: "None"
  END CROSS JOIN
RIGHT PLAN ON: []
  DF ["cat_1"]; PROJECT */1 COLUMNS; SELECTION: "None"
END CROSS JOIN  --- END STREAMING
```
Although everything is inside STREAMING block, `collect()` seems to be the bottleneck. I added a few lines to see where the process is killed for 10^10 search space size:

```
before collecting...
zsh: killed     python baybe/searchspace/test_polars_runtime.py
```
I pushed the updated code that uses `streaming=True` [here](https://github.com/emdgroup/baybe/pull/252/commits/e4d6b45211de324b61c019a83568c31a92d4bff2). It was used previously similar to the example above.

Reply:
I resolved this comment since we discussed it in our meeting.